### PR TITLE
DEMO Run 10

### PR DESCRIPTION
**Pull request by gondiaz**
This PR add DEMO Run 10 tracking plane configuration. The four modifications compared to Run 9 are the following:
- Internal sides of rectangular holes are TPB coated. A new seteable variable `hole_coated` is used.
- Dummy Xenon membranes are removed (setting their thickness to 0).
- Add opening holes in outer mask coating at SIPM positions.
- Add SiPM coating (still not implemented, waiting for a corresponding PR from @halmamol which adds this functionality in the Next100SiPM class).

### CODE REVIEW

File: source/geometries/NextDemoSiPMBoard.cc [hunk 1]
```diff
@@ -207,25 +209,59 @@ void NextDemoSiPMBoard::Construct()
...
+  G4VPhysicalVolume* hole_phys_ = new G4PVPlacement(sipm_rot, G4ThreeVector(0., 0., sipm_posz), sipm_->GetLogicalVolume(),
+                                                    sipm_->GetLogicalVolume()->GetName(), hole_logic,
+                                                    false, 0, false);
+
+  if (hole_coated && (hole_type_ == "rectangular")){
+
+    G4Box* hole_coating_top = new G4Box("HOLE_COATING", hole_x_/2., coating_thickn_/2., mask_thickn_/2.);
+    G4Box* hole_coating_lat = new G4Box("HOLE_COATING", coating_thickn_/2., hole_y_/2., mask_thickn_/2.);
+
+    G4UnionSolid* hole_coating = new G4UnionSolid("HOLE_COATING", hole_coating_top, hole_coating_lat, 0,
```
Comment (halmamol) [inline source/geometries/NextDemoSiPMBoard.cc:245]: I've thinking about it, I'm not fully sure if you're properly locating each of the sides at the proper location with respect to the others. What if instead of using `G4UnionSolid` to construct the volume, we use `G4SubtractionSolid`? We could define 2 boxes: one the size of the hole, and the other corresponding to the free volume after adding the coating. At the end you would only have the walls around the hole that correspond to the coated area. It would be much simpler.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 2]
```diff
@@ -127,12 +129,12 @@ void NextDemoSiPMBoard::Construct()
   G4Material* mother_gas = mother_phys_->GetLogicalVolume()->GetMaterial();
 
 
-  /// Board configuration checks
-  // Coating require membranes
-  if (coating_thickn_ > 0.)
-    if (membrane_thickn_ == 0.)
-      G4Exception("[NextDemoSiPMBoard]", "Construct()", FatalException,
-      "Coating require membranes");
+  // /// Board configuration checks
```
Comment (halmamol) [inline source/geometries/NextDemoSiPMBoard.cc]: Is there any reason of leaving these lines commented?

File: source/geometries/NextDemoSiPMBoard.cc (hunk 1 above)
Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.cc:245]: I agree that your solution would also be valid and probably easier to implement, and also the proper way to implement the coating for rounded holes. Nevertheless you can check that the sides are properly placed. In the figure below I show the coating placed inside the hole for a coating thickness of 1 mm (to improve visualization).

coated sipm hole

File: source/geometries/NextDemoSiPMBoard.cc (hunk 2 above)
Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.cc]: I had to comment the lines to make the code compatible with Run 10. Since coating is placed over the membranes in Run 7, Run 8 and Run 9, this exception might be useful. I commented out this lines just to remember this fact.

In my opinion, since the membranes we are simulating are dummy ones (just a Xenon volume with no effect in the simulation) I don't even think they are useful in the geometry. But that is something @jmunozv might want to clarify.

Comment (halmamol) [reply source/geometries/NextDemoSiPMBoard.cc]: I see. If need to keep them (depending on what experts say), we could add lines to do this check depending on the run configuration used. I guess this will mean to add a config variable.

File: source/geometries/NextDemoSiPMBoard.cc (hunk 1 above)
Comment (halmamol) [reply source/geometries/NextDemoSiPMBoard.cc:245]: Cool. I still think it would be more efficient and simple with the `G4SubtractionSolid`, it would be just two `G4Box` one inside the other. I'm not sure if you would have any issue in the joint areas.

Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.cc:245]: What issue? It is a `G4UnionSolid` thus the overlapping is managed. There is no geometry overlap either.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 3]
```diff
@@ -55,6 +56,7 @@ NextDemoSiPMBoard::NextDemoSiPMBoard():
   hole_diam_       (3.5  * mm),
   hole_x_          (0.0  * mm),
   hole_y_          (0.0  * mm),
+  hole_coated      (true),
```
Comment (halmamol) [inline source/geometries/NextDemoSiPMBoard.cc]: By convention this variable needs underscore.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 4]
```diff
@@ -259,32 +293,40 @@ void NextDemoSiPMBoard::Construct()
...
     new G4LogicalBorderSurface("TEFLON_WLS_GAS_OPSURF", coating_phys,
                                mother_phys_, coating_opsurf);
     new G4LogicalBorderSurface("GAS_TEFLON_WLS_OPSURF", mother_phys_,
                                coating_phys, coating_opsurf);
 
+    // Coating holes
+    if (membrane_thickn_ == 0.){
+      if (hole_type_ == "rectangular"){
+
+        G4Box* board_coating_hole_solid = new G4Box("BOARD_COATING_HOLE", hole_x_/2., hole_y_/2., coating_thickn_/2.);
```
Comment (halmamol) [inline source/geometries/NextDemoSiPMBoard.cc:327]: I found a little bit missleading the name **BOARD_COATING_HOLE**. I would add some lines specifying the purpouse of this section.

**Comment by halmamol** [comment]
I've realised that the `VERBOSITY` section of `NextDemoSiPMBoard.cc` it is not updated with the use of Next100SiPMs and rectangle mask holes. We could add these changes in this PR.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 5]
```diff
@@ -306,9 +352,18 @@ void NextDemoSiPMBoard::Construct()
 
     G4cout << "* Kapton thickness:   " << kapton_thickn_   << G4endl;
     G4cout << "* Mask thickness:     " << mask_thickn_     << G4endl;
-    G4cout << "* Mask hole diameter: " << hole_diam_       << G4endl;
     G4cout << "* Membrane thickness: " << membrane_thickn_ << G4endl;
-    G4cout << "* Coating thickness: "  << coating_thickn_  << G4endl;
+    G4cout << "* Coating thickness:  " << coating_thickn_  << G4endl;
+    G4cout << "* SiPM type:          " << sipm_type_       << G4endl;
```
Comment (halmamol) [inline source/geometries/NextDemoSiPMBoard.cc:362]: Maybe this line can be moved to line 346, or we could directly erase that one.

**Comment by paolafer** [comment]
The description if the `NextDemoTrackingPlane` class needs to be updated with the new configurations.

**Comment by paolafer** [comment]
The same applies to `NextDemoSiPMBoard`.

File: source/geometries/NextDemoSiPMBoard.cc (hunk 1 above)
Comment (paolafer) [reply source/geometries/NextDemoSiPMBoard.cc:245]: To move here the discussion in Slack: the best way to do it is putting a `G4Box` inside another one, no subtractions neither unions involved, which may slow down the tracking. The bigger box will be made of TPB and the smaller one of the gas of the mother volume. Also, if we ever need to do the same with the rounded holes, it can be done the same way.

Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.cc:245]: Since the SiPM resides in the hole, the two `G4Box` implementation you suggest would requiere to modify the volume were they are placed, and consequently the previous run configurations. The volumes are placed as intended and inner coating is only added if required, which  would not be the case with two boxes since one of them is made of tpb.

For external reader: we agreed via Slack discussion to maintain this implementation since is the minimal change possible. We would consider the change in the future if needed, for example, to implement rounded hole coatings.

File: source/geometries/NextDemoSiPMBoard.h [hunk 6]
```diff
@@ -1,11 +1,12 @@
 // -----------------------------------------------------------------------------
-// nexus | NextDemoSiPMBoard.h
+// nexus | NextDemoSiPMBoard.cc
```
Comment (paolafer) [inline source/geometries/NextDemoSiPMBoard.h]: Why this change? It should be `.h`.

File: source/geometries/NextDemoTrackingPlane.cc [hunk 7]
```diff
@@ -2,9 +2,10 @@
 // nexus | NextDemoTrackingPlane.cc
 //
 // Tracking plane of the Demo++ geometry.
-// It implements 2 different TP implementations:
+// It describes 2 different TP implementations:
```
Comment (paolafer) [inline source/geometries/NextDemoTrackingPlane.cc]: The implementations are 3, I think.

**Comment by paolafer** [comment]
Now we can add the `run10` configuration in the tests.

File: source/geometries/NextDemoSiPMBoard.h (hunk 6 above)
Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.h]: Rigth, I just copied the description in the .cc. Amended.

File: source/geometries/NextDemoTrackingPlane.cc (hunk 7 above)
Comment (gondiaz) [reply source/geometries/NextDemoTrackingPlane.cc]: Sure, amended.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 8]
```diff
@@ -208,25 +211,59 @@ void NextDemoSiPMBoard::Construct()
...
 
   // Generate SiPM positions
   GenerateSiPMPositions();
 
   G4double sipm_posz = - mask_thickn_/2. + sipm_z_dim/2.;
 
-  new G4PVPlacement(sipm_rot, G4ThreeVector(0., 0., sipm_posz), sipm_->GetLogicalVolume(),
-                    sipm_->GetLogicalVolume()->GetName(), hole_logic,
-                    false, 0, false);
+  G4VPhysicalVolume* hole_phys_ = new G4PVPlacement(sipm_rot, G4ThreeVector(0., 0., sipm_posz), sipm_->GetLogicalVolume(),
```
Comment (paolafer) [inline source/geometries/NextDemoSiPMBoard.cc]: Remove trailing underscore from the variable name in lines 235, 260 and 261.

Comment (paolafer) [inline source/geometries/NextDemoSiPMBoard.cc]: This placement is that of the SiPM inside the hole, right? If so it should be called `sipm_phys`, otherwise it's misleading.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 9]
```diff
@@ -208,25 +211,59 @@ void NextDemoSiPMBoard::Construct()
...
+    hole_coating = new G4UnionSolid("HOLE_COATING", hole_coating, hole_coating_top, 0,
+                                    G4ThreeVector(0., -hole_y_ + coating_thickn_, 0.));
+
+    G4LogicalVolume* hole_coating_logic = new G4LogicalVolume(hole_coating, tpb, "HOLE_COATING");
+
+    G4VPhysicalVolume* hole_coating_phys_ =
+          new G4PVPlacement(nullptr, G4ThreeVector(0., hole_y_/2. - coating_thickn_/2., 0.), hole_coating_logic,
+                            "HOLE_COATING", hole_logic, false, 0, false);
+
+    new G4LogicalBorderSurface("TEFLON_WLS_HOLE_GAS_OPSURF", hole_coating_phys_,
```
Comment (paolafer) [inline source/geometries/NextDemoSiPMBoard.cc]: If you use `hole_phys_` here, you're setting the surface between the SiPM and the coating, not between the gas of the hole and the coating, because `hole_phys_` is the physical volume of the SiPM in the current version of the code, right?

Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.cc:259]: The sipm volume is smaller than the hole volume. The hole has dimensions of the mask holes and sipms are placed inside. That means that most of the hole is filled with Xe, except the SiPM volume, in the  figure the solid box represents the full SiPM encasing and the coated walls:
SiPM-inside-holes

Comment (gondiaz) [reply source/geometries/NextDemoSiPMBoard.cc]: Right, `hole_phys` should be `sipm_phys`. These `G4LogicalBorderSurface` should be defined for each hole in its positioning at line 289, no?

Comment (paolafer) [reply source/geometries/NextDemoSiPMBoard.cc]: I would say so. However, I'm looking at our own documentation of the optical surfaces (https://github.com/next-exp/nexus/wiki/Setting-optical-properties) and there's a caveat regarding the case when one of the volumes is the daughter of the other one. I think we should try and find a time to talk to @jmunozv to clarify this.

File: source/geometries/NextDemoSiPMBoard.cc [hunk 10]
```diff
@@ -250,42 +281,62 @@ void NextDemoSiPMBoard::Construct()
 
 
   /// Placing the Holes with SiPMs & membranes inside
+  G4VPhysicalVolume* hole_phys[num_sipms_];
```
Comment (paolafer) [inline source/geometries/NextDemoSiPMBoard.cc]: It looks like this gives a compilation warning: `variable length arrays are a C99 feature`.
You could use an `std::vector` instead.

Comment (paolafer) [reply source/geometries/NextDemoSiPMBoard.cc]: After talking to @jmalbos, it seems that there is no need to define any container. You can simply assign the same pointer to the placement in each iteration of the loop.

File: source/geometries/NextDemoSiPMBoard.cc (hunk 2 above)
Comment (halmamol) [reply source/geometries/NextDemoSiPMBoard.cc]: What are we going to do with these commented lines? Are we going to keep them like that or adapt them to the case without membrane? Maybe we could include `hole_coated_` in the condition, like `if (!hole_coated_ && (membrane_thickn_ == 0.) )`

**Comment by halmamol** [comment]
This PR adds Run 10 in DEMO configuration. This run removes the membranes, and includes the possibility to coat with TPB the holes in the mask. In each of these holes, optical surfaces have been assigned to coating. It also sets the coating of the SiPMs from the Next100SiPM class. Tests have been included accordingly for Run 10. I approve this PR, but a final blessing would be expected from more experienced reviewers (@paolafer @jmalbos @jmunozv)

**Review by halmamol — APPROVED** [review]
This PR adds Run 10 in DEMO configuration. This run removes the membranes, and includes the possibility to coat with TPB the holes in the mask. In each of these holes, optical surfaces have been assigned to coating. It also sets the coating of the SiPMs from the Next100SiPM class. Tests have been included accordingly for Run 10. Future studies about the optical properties of daughter volumes would be required in order to improve the implementation of the coating of the holes. I approve this PR, but I would wait for a final blessing from more experienced reviewers (@paolafer @jmalbos @jmunozv)
